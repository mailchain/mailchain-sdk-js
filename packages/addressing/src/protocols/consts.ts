// WARNING: CAREFUL WHEN CHANGING THE VALUES OF THIS CONSTANTS, THEY MIGHT BE ALREADY LINKED TO KEY PATHS

/**
 * Algorand protocol name.
 */
export const ALGORAND = 'algorand' as const;
/**
 * Ethereum protocol name.
 */
export const ETHEREUM = 'ethereum' as const;
/**
 * Substrate protocol name.
 */
export const SUBSTRATE = 'substrate' as const;

export const NEAR = 'near' as const;

export const TEZOS = 'tezos' as const;

export const FILECOIN = 'filecoin' as const;

/**
 * Mailchain protocol name.
 */
export const MAILCHAIN = 'mailchain' as const;

export const ALL_PROTOCOLS = [ALGORAND, ETHEREUM, SUBSTRATE, NEAR, MAILCHAIN, TEZOS, FILECOIN] as const;
export type ProtocolType = (typeof ALL_PROTOCOLS)[number];

const ENABLED_PROTOCOLS = [ETHEREUM, NEAR, TEZOS, FILECOIN] as const;
export type EnabledBlockchainProtocol = (typeof ENABLED_PROTOCOLS)[number];
export function isBlockchainProtocolEnabled(protocol: string): protocol is EnabledBlockchainProtocol {
	return ENABLED_PROTOCOLS.includes(protocol as EnabledBlockchainProtocol);
}
