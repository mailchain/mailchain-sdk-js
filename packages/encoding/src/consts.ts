/**
 * Plain hex encoding value.
 */
export const HEX = 'hex/plain';

/**
 * 0x prefixed hex encoding
 */
export const HEX_0X_PREFIX = 'hex/0x-prefix';

export const BASE32 = 'base32/plain';
export const BASE58 = 'base58/plain';
export const BASE64 = 'base64/plain';

export type Encodings = typeof HEX | typeof HEX_0X_PREFIX | typeof BASE32 | typeof BASE58 | typeof BASE64;
