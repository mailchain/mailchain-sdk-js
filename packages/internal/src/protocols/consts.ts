/**
 * Algorand protocol name.
 */
export const Algorand = 'algorand';
/**
 * Ethereum protocol name.
 */
export const Ethereum = 'ethereum';
/**
 * Substrate protocol name.
 */
export const Substrate = 'substrate';
