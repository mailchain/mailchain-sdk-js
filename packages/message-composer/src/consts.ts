/** Recommended line length limit defined in https://www.rfc-editor.org/rfc/rfc5322#section-2.1.1 */
export const LINE_LENGTH_FOLD = 78;

export const CRLF = '\r\n' as const;

export const HEADER_LABELS = {
	MimeVersion: 'MIME-Version',
	MessageId: 'Message-ID',
	Subject: 'Subject',
	Date: 'Date',
	From: 'From',
	To: 'To',
	Cc: 'Cc',
	Bcc: 'Bcc',
	ContentType: 'Content-Type',
	ContentTransferEncoding: 'Content-Transfer-Encoding',
	ContentDisposition: 'Content-Disposition',
	ContentId: 'Content-ID',
} as const;
