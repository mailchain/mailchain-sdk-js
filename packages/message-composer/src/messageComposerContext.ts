import { encodeBase64, encodeBase64UrlSafe } from '@mailchain/encoding/base64';
import { decodeUtf8 } from '@mailchain/encoding/utf8';

export type MessageComposerContext = {
	random: (len: number) => Uint8Array;
	decodeUtf8: (content: string) => Promise<Uint8Array>;
	encodeBase64: (content: Uint8Array, urlSafe?: boolean) => Promise<string>;
};

export function defaultMessageComposerContext(): MessageComposerContext {
	return {
		random: (len) => crypto.getRandomValues(new Uint8Array(len)),
		decodeUtf8: (content) => Promise.resolve(decodeUtf8(content)),
		encodeBase64: (content, urlSafe) =>
			urlSafe ? Promise.resolve(encodeBase64UrlSafe(content)) : Promise.resolve(encodeBase64(content)),
	};
}
