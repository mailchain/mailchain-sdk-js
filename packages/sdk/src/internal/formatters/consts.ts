export const X_IDENTITY_KEYS = 'X-IdentityKeys' as const;
