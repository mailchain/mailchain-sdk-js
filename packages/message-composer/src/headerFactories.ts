import { HEADER_LABELS } from './consts';
import { Header } from './types';

export function createHeader<T>(label: string, value: T, attrs?: Header<any>['attrs']): Header<T> {
	return { label, value, attrs };
}

export function contentTypeBoundaryHeader(type: 'alternative' | 'mixed', boundary: string): Header<string> {
	return createHeader(HEADER_LABELS.ContentType, `multipart/${type}`, [['boundary', boundary]]);
}
