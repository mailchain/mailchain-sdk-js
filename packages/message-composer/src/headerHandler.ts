import formatDate from 'date-fns/format';
import enUS from 'date-fns/locale/en-US';
import { hasOnlyPrintableUsAscii } from './hasOnlyAscii';
import { semanticLineFold, simpleHardFold } from './folding';
import {
	AddressHeader,
	DateHeader,
	Header,
	isAddressHeader,
	isDateHeader,
	isStringHeader,
	StringHeader,
} from './types';
import { CRLF, LINE_LENGTH_FOLD } from './consts';
import { MessageComposerContext } from './messageComposerContext';

/**
 * Processes all the headers and outputs single `string` chunk that represents the headers into MIME format.
 *
 * Note: there isn't leading/trailing CRLF of the output, you need to add your own if you want one.
 */
export async function buildHeaders(headers: Header<any>[], ctx: MessageComposerContext): Promise<string> {
	const exportedHeaders = new Array<string>();
	for (const header of headers) {
		const exportedHeader = await exportHeader(header, ctx);
		exportedHeaders.push(exportedHeader);
	}
	return exportedHeaders.join(CRLF);
}

/**
 * Export single {@link Header} into it's MIME format. May contain multiple lines if the contents have been folded.
 *
 * Example: `"Subject: This is the subject of the message"`
 */
export async function exportHeader(header: Header<any>, ctx: MessageComposerContext): Promise<string> {
	const headerValue = await exportHeaderValue(header, ctx);
	if (header.attrs == null || header.attrs.length === 0) return `${header.label}: ${headerValue}`;

	const headerAttrs = await exportHeaderAttributes(header.attrs, ctx);

	// TODO: naive join of attrs, just appending at the end. Need to take care of length
	return [`${header.label}: ${headerValue}`, ...headerAttrs].join('; ');
}

/** Export just the {@link Header#value} part of the header. The {@link Header#label} is not exported. */
export async function exportHeaderValue(header: Header<any>, ctx: MessageComposerContext): Promise<string> {
	if (isStringHeader(header)) return exportStringHeader(header, ctx);
	if (isDateHeader(header)) return exportDateHeader(header, ctx);
	if (isAddressHeader(header)) return exportAddressHeader(header, ctx);

	throw new Error(`cannot export value for header [${header.label}]`);
}

/** Export the {@link Header#attrs} part of the header as separate element in the resulting `string[]` array. */
export async function exportHeaderAttributes(
	attrs: NonNullable<Header<any>['attrs']>,
	ctx: MessageComposerContext,
): Promise<string[]> {
	// TODO: Not taking line length limit into consideration
	return await Promise.all(
		attrs.map(async (attr) => {
			const value = hasOnlyPrintableUsAscii(attr[1])
				? attr[1]
				: `=?UTF-8?B?${await ctx.encodeBase64(await ctx.decodeUtf8(attr[1]))}?=`;

			if (attr[0] != null) return `${attr[0]}="${value}"`;
			return `"${value}"`;
		}),
	);
}

/** Export the `string` value out of {@link StringHeader}. It takes into consideration length of the value and the encoding of it. */
export async function exportStringHeader(header: StringHeader, ctx: MessageComposerContext): Promise<string> {
	if (hasOnlyPrintableUsAscii(header.value)) {
		return semanticLineFold(header.value, LINE_LENGTH_FOLD);
	}

	const utf8Array = await ctx.decodeUtf8(header.value);
	const base64Encoded = await ctx.encodeBase64(utf8Array);
	return simpleHardFold(base64Encoded, {
		lineLength: LINE_LENGTH_FOLD,
		prefix: '=?UTF-8?B?',
		suffix: '?=',
		padNewLines: true,
		encodingConsideration: 'base64',
	});
}

/**
 * Export the `Date` value out of {@link DateHeader}.
 *
 * Example: Wed, 17 Aug 2022 10:01:56 +0000
 */
export async function exportDateHeader(header: DateHeader, ctx: MessageComposerContext): Promise<string> {
	return formatDate(header.value, 'EEE, dd MMM yyyy HH:mm:ss xxxx', { locale: enUS });
}

/**
 * Export the addresses out of the {@link AddressHeader}.
 * It joins every address into new line divided with `",CRLF "`, ready to be appended to the header label.
 *
 * Example:
 * ```
 * "Alice Lastname" <alice@mailchain.com>,
 *  "Bob Lastname" <bob@mailchain.com>
 * ```
 */
export async function exportAddressHeader(header: AddressHeader, ctx: MessageComposerContext): Promise<string> {
	// TODO: Check for other ASCII characters that need to be escaped
	// TODO: Check for non ASCII characters (base64 encode)
	// TODO: check for length (going optimistically for now, splitting on each recipient)
	return header.value
		.map((r) => {
			return Boolean(r.name?.length) ? `"${r.name}" <${r.address}>` : `<${r.address}>`;
		})
		.join(`,${CRLF} `);
}
