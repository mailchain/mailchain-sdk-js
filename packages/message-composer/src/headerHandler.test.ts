import { CRLF, HTAB, LINE_LENGTH_FOLD } from './consts';
import { createHeader } from './headerFactories';
import { exportHeader, exportHeaderAttributes, exportStringHeader } from './headerHandler';
import { defaultMessageComposerContext } from './messageComposerContext';

describe('headerHandler', () => {
	describe('exportHeader', () => {
		it('should export string header', async () => {
			const header = createHeader('Label', 'stringValue', ['key', 'value']);

			const res = await exportHeader(header, defaultMessageComposerContext());

			expect(res).toEqual(`Label: stringValue; key="value"`);
		});

		it('should export header with empty value', async () => {
			const header = createHeader('Label', '', ['key', 'value']);

			const res = await exportHeader(header, defaultMessageComposerContext());

			expect(res).toEqual(`Label: key="value"`);
		});

		it('should export date header', async () => {
			const header = createHeader('Label', new Date('01/01/2001'), ['key', 'value']);

			const res = await exportHeader(header, defaultMessageComposerContext());

			expect(res).toEqual(`Label: Mon, 01 Jan 2001 00:00:00 +0000; key="value"`);
		});

		it('should export single address header', async () => {
			const header = createHeader('Label', [{ name: 'Alice', address: 'alice@mailchain.com' }]);

			const res = await exportHeader(header, defaultMessageComposerContext());

			expect(res).toEqual(`Label: "Alice" <alice@mailchain.com>`);
		});

		it('should export multiple address header', async () => {
			const header = createHeader('Label', [
				{ name: 'Alice', address: 'alice@mailchain.co' },
				{ name: 'Bob', address: 'bob@mailchain.co' },
				{ name: 'Rob', address: 'rob@mailchain.xyz' },
			]);

			const res = await exportHeader(header, defaultMessageComposerContext());

			expect(res).toEqual(
				`Label: "Alice" <alice@mailchain.co>,${CRLF} "Bob" <bob@mailchain.co>,${CRLF} "Rob" <rob@mailchain.xyz>`,
			);
		});

		it('should export header with multiple attributes not being folded', async () => {
			const header = createHeader('Label', 'value', ['key1', 'value1'], ['key2', 'value2'], ['key3', 'value3']);

			const res = await exportHeader(header, defaultMessageComposerContext());

			expect(res).toEqual(`Label: value; key1="value1"; key2="value2"; key3="value3"`);
		});

		it('should export header with multiple attributes with attribute folded', async () => {
			const header = createHeader(
				'Label',
				'value',
				['attr-key-1', 'long-value-1'],
				['attr-key-2', 'extra-long-value-2'],
				['attr-key-3', 'very-very-very-very-very-very-long-value-that-will-get-folded-3'],
			);

			const res = await exportHeader(header, defaultMessageComposerContext());

			expect(res).toEqual(
				`Label: value; attr-key-1="long-value-1"; attr-key-2="extra-long-value-2";${CRLF}${HTAB}attr-key-3="very-very-very-very-very-very-long-value-that-will-get-folded-3"`,
			);
		});
	});

	describe('exportHeaderAttributes', () => {
		it('should export single attributes', async () => {
			const res = await exportHeaderAttributes([['key', 'value']], defaultMessageComposerContext());

			expect(res).toEqual([`key="value"`]);
		});

		it('should export multiple attributes', async () => {
			const res = await exportHeaderAttributes(
				[
					['key1', 'value1'],
					['key2', 'value2'],
				],
				defaultMessageComposerContext(),
			);

			expect(res).toEqual([`key1="value1"`, `key2="value2"`]);
		});

		it('should export attribute with non US-ASCII char value', async () => {
			const res = await exportHeaderAttributes([['key', 'валуе 🥸']], defaultMessageComposerContext());

			expect(res).toEqual([`key="=?UTF-8?B?0LLQsNC70YPQtSDwn6W4?="`]);
		});

		it('should expert attribute without key', async () => {
			const res = await exportHeaderAttributes([[undefined, 'value']], defaultMessageComposerContext());

			expect(res).toEqual([`"value"`]);
		});
	});

	describe('exportStringHeader', () => {
		it('should export simple string value', async () => {
			const res = await exportStringHeader(createHeader('', 'value'), defaultMessageComposerContext());

			expect(res).toEqual('value');
		});

		it('should semanticaly fold US-ASCII only line', async () => {
			const res = await exportStringHeader(
				createHeader(
					'Label',
					'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Elit ullamcorper dignissim cras tincidunt lobortis feugiat vivamus at. Blandit aliquam etiam erat velit scelerisque in dictum non. Sed viverra tellus in hac habitasse platea dictumst vestibulum. Nibh tellus molestie nunc non. Id leo in vitae turpis massa sed elementum tempus. Ante metus dictum at tempor commodo ullamcorper a lacus vestibulum. Non odio euismod lacinia at quis risus sed. Imperdiet massa tincidunt nunc pulvinar. Est ante in nibh mauris. Cursus euismod quis viverra nibh cras. Duis at tellus at urna condimentum mattis pellentesque. Amet mauris commodo quis imperdiet massa tincidunt nunc. Vitae congue mauris rhoncus aenean vel elit scelerisque mauris.',
				),
				defaultMessageComposerContext(),
			);

			res.split(CRLF).forEach((line, index) => {
				expect(line.length).toBeLessThan(LINE_LENGTH_FOLD);
				if (index > 0) {
					// every line begins with white space char
					expect(line.match(/$\s/)).toBeDefined();
				}
			});
			expect(res).toMatchSnapshot();
		});

		it('should export as UTF-8 base64 encoded', async () => {
			const res = await exportStringHeader(
				createHeader(
					'Label',
					'途回福構野市天時番児事日期保。談真望載最学書表殺祉再特後日路要人。議渡小真設俳落権数左広員意竹吸更地言。択割特図無横原油集目再晴大報用左産。活必名引大低普浜告見窓幌験井続化。世供芸負習学授戦質左改鈴述禁賠能薄。関超主点切世治磨発車議禁人政。究撃横了訴話者育的強天展調九旭撃能済。能皇臓哲天権目鹿市全求各守迎細台樸技分',
				),
				defaultMessageComposerContext(),
			);

			res.split(CRLF).forEach((line, index) => {
				expect(line.length).toBeLessThan(LINE_LENGTH_FOLD);
				line.trim().startsWith('=?UTF-8?B?');
				line.endsWith('?=');
				if (index > 0) {
					// every line begins with white space char
					expect(line.match(/$\s/)).toBeDefined();
				}
			});
			expect(res).toMatchSnapshot();
		});
	});
});
