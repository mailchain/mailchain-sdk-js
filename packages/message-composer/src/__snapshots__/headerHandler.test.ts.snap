// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`headerHandler exportStringHeader should export as UTF-8 base64 encoded 1`] = `
"=?UTF-8?B?6YCU5Zue56aP5qeL6YeO5biC5aSp5pmC55Wq5YWQ5LqL5pel5pyf5L+d44CC6KuH?=
	=?UTF-8?B?55yf5pyb6LyJ5pyA5a2m5pu46KGo5q6656WJ5YaN54m55b6M5pel6Lev6KaB5Lq6?=
	=?UTF-8?B?44CC6K2w5rih5bCP55yf6Kit5L+z6JC95qip5pWw5bem5bqD5ZOh5oSP56u55ZC4?=
	=?UTF-8?B?5pu05Zyw6KiA44CC5oqe5Ymy54m55Zuz54Sh5qiq5Y6f5rK56ZuG55uu5YaN5pm0?=
	=?UTF-8?B?5aSn5aCx55So5bem55Sj44CC5rS75b+F5ZCN5byV5aSn5L2O5pmu5rWc5ZGK6KaL?=
	=?UTF-8?B?56qT5bmM6aiT5LqV57aa5YyW44CC5LiW5L6b6Iq46LKg57+S5a2m5o6I5oim6LOq?=
	=?UTF-8?B?5bem5pS56Yi06L+w56aB6LOg6IO96JaE44CC6Zai6LaF5Li754K55YiH5LiW5rK7?=
	=?UTF-8?B?56Oo55m66LuK6K2w56aB5Lq65pS/44CC56m25pKD5qiq5LqG6Ki06Kmx6ICF6IKy?=
	=?UTF-8?B?55qE5by35aSp5bGV6Kq/5Lmd5pet5pKD6IO95riI44CC6IO955qH6IeT5ZOy5aSp?=
	=?UTF-8?B?5qip55uu6bm/5biC5YWo5rGC5ZCE5a6I6L+O57Sw5Y+w5qi45oqA5YiG?="
`;

exports[`headerHandler exportStringHeader should semanticaly fold US-ASCII only line 1`] = `
"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
 tempor incididunt ut labore et dolore magna aliqua. Elit ullamcorper
 dignissim cras tincidunt lobortis feugiat vivamus at. Blandit aliquam etiam
 erat velit scelerisque in dictum non. Sed viverra tellus in hac habitasse
 platea dictumst vestibulum. Nibh tellus molestie nunc non. Id leo in vitae
 turpis massa sed elementum tempus. Ante metus dictum at tempor commodo
 ullamcorper a lacus vestibulum. Non odio euismod lacinia at quis risus sed.
 Imperdiet massa tincidunt nunc pulvinar. Est ante in nibh mauris. Cursus
 euismod quis viverra nibh cras. Duis at tellus at urna condimentum mattis
 pellentesque. Amet mauris commodo quis imperdiet massa tincidunt nunc. Vitae
 congue mauris rhoncus aenean vel elit scelerisque mauris."
`;
